### PR TITLE
Add a ``write-started`` event

### PR DESCRIPTION
**Pull request by chrisjsewell**
Currently, there is no event that triggers consistently just before the WRITE phase starts.

This is necessary for any extensions that need to perform any "global setup" (for all builders),
just before post-transforming documents, but with information of the fully written `BuildEnvironment`.

In particular, this is desperately needed for https://github.com/useblocks/sphinx-needs,
where they are currently having to do some horrible workarounds.

The updated diagram in https://sphinx--12567.org.readthedocs.build/en/12567/extdev/event_callbacks.html explains where this fits in:

image

### CODE REVIEW

File: sphinx/events.py [hunk 1]
```diff
@@ -45,6 +45,7 @@ class EventListener(NamedTuple):
     'warn-missing-reference': 'domain, node',
     'doctree-resolved': 'doctree, docname',
     'env-updated': 'env',
+    'write-starting': 'builder',
```
Comment (picnixz) [inline sphinx/events.py]: To be consistent with the other names, I would have chosen `write-started` or `builder-write`

File: doc/extdev/event_callbacks.rst [hunk 2]
```diff
@@ -52,21 +52,25 @@ Below is an overview of the core event that happens during a build.
 
    10. event.env-updated(app, env)
    11. event.env-get-updated(app, env)
-   12. event.env-check-consistency(app, env)
+
+   if environment is written to disk:
+      12. event.env-check-consistency(app, env)
+
+   13. event.write-starting(app, builder)
```
Comment (picnixz) [inline doc/extdev/event_callbacks.rst]: Maybe add a comment saying that this is done after determining whether to write in parallel or serial mode (i.e., `app.parallel_ok` is set correctly and should not be changed at that time).

Comment (chrisjsewell) [reply doc/extdev/event_callbacks.rst]: done

File: sphinx/events.py (hunk 1 above)
Comment (chrisjsewell) [reply sphinx/events.py]: changed to `write-started`

**Review by AA-Turner — APPROVED** [review]

**Pull request merged**